### PR TITLE
Estimation engine/api

### PR DESCRIPTION
**Pull request by sorja**
- [x] Pass generateSpec in correct format from frontend to backend
- [x] Update database with generated data
- [x] Return generated data back to frontend, update

### CODE REVIEW

File: client/pages/AssessmentSection/DataTable/GenerateValues/useGenerateValules.ts [hunk 1]
```diff
@@ -66,20 +74,23 @@ const useGenerateValues = (
...
+          tableName,
+          fields: fields
+            .filter((f) => f.selected)
+            .map(({ annualChangeRates, variableName }) => ({
+              annualChangeRates,
+              variableName,
+              changeRates,
+            })),
+        })
+      )
```
Comment (sorja) [inline client/pages/AssessmentSection/DataTable/GenerateValues/useGenerateValules.ts:93]: this

File: client/store/pages/assessmentSection/actions/postEstimate.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        params: {
+          countryIso,
+          assessmentName,
+          cycleName,
+          method,
+          sectionName,
+          tableName,
+        },
+      }
+    )
```
Comment (sorja) [inline client/store/pages/assessmentSection/actions/postEstimate.ts:37]: this

File: server/api/assessment/postEstimation.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    const { assessment, cycle } = await AssessmentController.getOneWithCycle({ name: assessmentName, cycleName })
+
+    const tableSpec = await TableRepository.getOne({ assessment, cycle, tableName })
+    const originalDataPointValues = await DataRepository.getOriginalDataPointData({
+      countryIso: countryIso as CountryIso,
+      cycle,
+      assessment,
+    })
+
+    const years = tableSpec.props.columnNames.map((column: string) => Number(column))
```
Comment (sorja) [inline server/api/assessment/postEstimation.ts:31]: tableSpec.props.columnNames[cycleUuid]

File: server/api/assessment/index.ts [hunk 4]
```diff
@@ -23,6 +24,9 @@ export const AssessmentApi = {
     express.get(ApiEndPoint.Assessment.country(), AuthMiddleware.requireView, getCountry)
     express.post(ApiEndPoint.Assessment.country(), AuthMiddleware.requireEdit, postCountry)
 
+    // Estimation
+    express.post(ApiEndPoint.Assessment.TableData.Estimate.many(), /* AuthMiddleware.requireEdit, */ postEstimation)
```
Comment (minotogna) [inline server/api/assessment/index.ts:28]: the AuthMiddleware.requireEdit should be enabled

File: server/api/assessment/postEstimation.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  try {
+    const { countryIso, assessmentName, cycleName, sectionName, method, tableName } = <Record<string, string>>req.query
+    const { fields } = <
+      {
+        fields: Array<{
+          annualChangeRates: { past: string; future: string }
+          variableName: string
+        }>
+      }
+    >req.body
```
Comment (minotogna) [inline server/api/assessment/postEstimation.ts:22]: why not to pass all in body ?

File: server/api/assessment/postEstimation.ts [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+      cycleName,
+      metaCache: true,
+    })
+
+    const tableSpec = await TableRepository.getOne({ assessment, cycle, tableName })
+    const originalDataPointValues = await DataRepository.getOriginalDataPointData({
+      countryISOs: [countryIso as CountryIso],
+      cycle,
+      assessment,
+    })
```
Comment (minotogna) [inline server/api/assessment/postEstimation.ts:34]: apis should use controllers

File: server/controller/cycleData/persistNodeValues.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+// Wrapper to support persisting full tables
+export const persistNodeValues = async (props: Props): Promise<void> => {
+  const { assessment, cycle, user, tableData, tableName } = props
+  const promises = Object.entries(tableData).flatMap(([countryIso, countryValues]) =>
+    Object.entries(countryValues).flatMap(([_section, sectionValues]) =>
+      Object.entries(sectionValues).flatMap(([colName, colValues]) =>
+        Object.entries(colValues).flatMap(([variableName, value]) => {
+          return persistNodeValue({
```
Comment (minotogna) [inline server/controller/cycleData/persistNodeValues.ts:23]: I'm wondering if there's a way to remove all this iterations.
what if the estimation engine returns an array of nodeUpdates e.g.
 ```typescript
[
{countryIso, colName, tableName, assessment, cycle, variableName, value},
...
]
```
 ?
 This way, the returned array is input to this persistNodeValues method and Props would become:
 ```typescript
 interface Props {
   nodes: Array<NodeUpdate>
   user: User
 }
 ```

**Review by minotogna — APPROVED** [review]

**Pull request merged**